velox/exec/tests/WindowTest.cpp

```diff
@@ -79,6 +80,136 @@ TEST_F(WindowTest, spill) {
   ASSERT_GT(stats.spilledPartitions, 0);
 }
 
+TEST_F(WindowTest, rankWithEqualValue) {
```

Comment:
@JkSelf @aditi-pandit Do we have a test that OOMs without the changes and succeeds with the changes and illustrates that the new logic is using less memory?

Reply:
@mbasmanova : No. I don't think such a test has been added. The tests ensure that we trigger and validate the rowsStreamingBuild paths.

Reply:
Should we add such a tests? After all, the only point of adding this new code path is to allow use cases which used to OOM to succeed without providing them more memory, no?

Reply:
@mbasmanova : I agree the test is ideal. I haven't researched the Velox test infrastructure enough to have a design for such a test. Though it seemed more feasible for the author to add on the Spark side as the memory arbitration in Gluten is independent of Velox memory arbitration I had presumed.

Reply:
@mbasmanova @aditi-pandit 
To conduct OOM  testing, we need to set up specific scenarios to effectively trigger memory overflow. Currently, we have added tests for Windows OOM on our Gluten CI . With this PR , we can also address the Windows OOM cases reported by our customers. There are three WindowBuild (Sort, Streaming, RowsStreaming), which essentially differ only in the way WindowPartitions are created. Therefore, I believe that as long as we ensure that all code paths are tested in our unit tests, that should suffice.

Reply:
> To conduct OOM testing, we need to set up specific scenarios to effectively trigger memory overflow.

This is true, but is also fairly straightforward to do. It is important to have standalone tests in Velox for key functionality. Otherwise, this functionality can break easily.

CC: @xiaoxmeng

Reply:
@xiaoxmeng Do you have any inputs? Thanks.

Reply:
@mbasmanova @aditi-pandit @rui-mo @xiaoxmeng 
Added one unit test to compare the peak memory usage in SortBasedWindowBuild and RowBasedStreamingWindow. Can you help to review again? Thanks.

Reply:
The test looks good to me. @mbasmanova Would you like to take a review? Thanks!

Reply:
@rui-mo @JkSelf Thank you for adding a test. Would you explain how the test works? Does the test fail with an OOM before the change? I'm out next week, but @xiaoxmeng is back.

Reply:
@mbasmanova @xiaoxmeng No, this test did not trigger an OOM. I tried adjusting the memory pool's capacity to trigger an OOM, but since a sort operator needs to be added before the streaming window, which also needs to hold the entire partition's data, the memory usage of the sort window and the streaming window is the same. The unit test I've added now only compares the peak memory usage of the Window operator. Since the SortWindow needs to hold the entire partition's data, its peak memory usage will increase as the volume of data grows. However, the RowBased window only needs to hold the data of one batch in a partition, so its peak memory usage will tend to stabilize. Here we compared two different datasets of 100,000 and 1,000,000. For the sort-based window, memory usage increases with the size of the dataset (2418624 vs 17098688). For the Row-based window, since different datasets hold the data of one batch, the memory usage is the same.

Reply:
@JkSelf 

> a sort operator needs to be added before the streaming window, which also needs to hold the entire partition's data, the memory usage of the sort window and the streaming window is the same. 

Oh... I wish you shared this challenge earlier. There is a simple solution. Make a custom test-only plan node and operator that generates sorted data on the fly. Say you tell it how many batches to produce and how many rows to put in a batch and it does that without using much memory (only what's needed for a single batch). Then push window on top, then count(1) aggregation. This way all the memory usage will be in the window and original implementation would OOMs, but streaming one would succeed. 

CC: @xiaoxmeng

Reply:
@JkSelf The purpose of the UT is to prevent the future modification breaking the streaming window's functionality. So we can configure a suitable threshold that sort can cache the data without OOM, but once streaming window cache the data, OOM should be triggered, and UT fails. Can we do this?

Reply:
@mbasmanova @FelixYBW Thanks for your suggestions. I added the OOM unit test with RowStreamingWindow [here](https://github.com/facebookincubator/velox/pull/9025/commits/7e77104d429f663297f53b4bc731a73659798d94). 
@xiaoxmeng Can you help to review if you have time? Thanks for your help.